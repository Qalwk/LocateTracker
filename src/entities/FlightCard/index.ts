export { FlightCard } from './ui/FlightCard';
