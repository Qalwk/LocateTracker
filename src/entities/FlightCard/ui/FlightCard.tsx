import clsx from "clsx";
import { Heart } from "lucide-react";

import React, { forwardRef } from "react";
import { useLocation, useNavigate } from "react-router";

import FlightStatus from "entities/Flight/ui/FlightStatus";

import { type Flight } from "shared/mocks/FlightsData";

import styles from "./FlightCard.module.scss";

interface FlightCardProps {
  flight: Flight;
  onClick?: (flight: Flight) => void;
  active?: boolean;
  isFavorite: boolean;
  onLikeClick: () => void;
  progress: number;
  ref?: React.RefObject<HTMLDivElement>;
}

function slugify(str: string) {
  return str.toLowerCase().replace(/\s+/g, "-");
}

export const FlightCard = forwardRef<HTMLDivElement, FlightCardProps>(
  function FlightCard(
    { flight, onClick, active = false, isFavorite, onLikeClick, progress },
    ref,
  ) {
    const navigate = useNavigate();
    const location = useLocation();

    const handleLinked = () => {
      const airline = slugify(flight.airline || "unknown");
      const from = slugify(flight.from?.city || "unknown");
      const to = slugify(flight.to?.city || "unknown");
      if (location.pathname.includes("/favorites")) {
        navigate(`/favorites/${airline}/${from}-${to}`);
      } else {
        navigate(`/${airline}/${from}-${to}`);
      }
      if (onClick) onClick(flight);
    };

    return (
      <div
        ref={ref}
        onClick={handleLinked}
        className={clsx(styles.flightCard, { [styles.active]: active })}
      >
        <div className={styles.header}>
          <div className={styles.airline}>
            <div className={styles.airlineLogo}>
              <img
                // src={flight.airlineLogo}
                src={`http://localhost:3001/images/${flight.airlineLogo || "default-logo.png"}`}
                alt="airlineLogo"
                onError={(e) => {
                  e.currentTarget.src =
                    "http://localhost:3001/images/default-logo.png";
                }}
              />
            </div>
            <div className={styles.flightNumber}>
              {flight.flightNumber || flight.id}
            </div>
          </div>
          <div className={styles.codes}>
            <span className={styles.code}>{flight.codes?.[0] || ""}</span>
            <span className={`${styles.code} ${styles.codeSecond}`}>
              {flight.codes?.[1] || ""}
            </span>
            <button
              className={clsx(styles.btnLike, {
                [styles.btnLikeActive]: isFavorite,
              })}
              onClick={(e) => {
                e.stopPropagation();
                onLikeClick();
              }}
            >
              <Heart size={16} fill={isFavorite ? "currentColor" : "none"} />
            </button>
          </div>
        </div>
        <div className={styles.airports}>
          <div className={`${styles.airport} ${styles.from}`}>
            <span className={styles.city}>
              {flight.from?.city || "Unknown"}
            </span>
            <span className={styles.iata}>{flight.from?.iata || "N/A"}</span>
          </div>
          <FlightStatus progress={progress} />
          <div className={`${styles.airport} ${styles.to}`}>
            <span className={styles.city}>{flight.to?.city || "Unknown"}</span>
            <span className={styles.iata}>{flight.to?.iata || "N/A"}</span>
          </div>
        </div>
      </div>
    );
  },
);
