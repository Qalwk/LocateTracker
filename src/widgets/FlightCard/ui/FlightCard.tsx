import { useNavigate } from "react-router";
import { type Flight } from "shared/mocks/FlightsData"
import styles from './FlightCard.module.scss'
import clsx from "clsx";
import { Heart } from "lucide-react";
import FlightStatus from "entities/Flight/ui/FlightStatus";

interface FlightCardProps {
  flight: Flight;
  onClick?: (flight: Flight) => void;
  active?: boolean;
  isFavorite: boolean;
  onLikeClick: () => void;
  progress: number;
}

export function FlightCard({
  flight,
  onClick,
  active = false,
  isFavorite,
  onLikeClick,
  progress,
}: FlightCardProps) {
  
  const navigate = useNavigate();

  function slugify(str: string) {
    return str.toLowerCase().replace(/\s+/g, '-');
  }

  const handleLinked = () => {
    const airline = slugify(flight.airline);
    const from = slugify(flight.from.city);
    const to = slugify(flight.to.city);
    navigate(`/${airline}/${from}-${to}`);
    if (onClick) onClick(flight);
  };

  return (
    <div 
      onClick={handleLinked}
      className={clsx(styles.flightCard, { [styles.active]: active })}
    >
      <div className={styles.header}>
        <div className={styles.airline}>
          <div className={styles.airlineLogo}>
              <img src={flight.airlineLogo} alt="airlineLogo" />
          </div>
          <div className={styles.flightNumber}>{flight.flightNumber}</div>
        </div>
        <div className={styles.codes}>
          <span className={styles.code}>{flight.codes[0]}</span>
          <span className={styles.code}>{flight.codes[1]}</span>
          <button
            className={clsx(styles.btnLike, { [styles.btnLikeActive]: isFavorite })}
            onClick={e => { e.stopPropagation(); onLikeClick(); }}
          >
            <Heart size={16} fill={isFavorite ? "currentColor" : "none"} />
          </button>
        </div>
      </div>
      <div className={styles.airports}>
        <div className={`${styles.airport} ${styles.from}`}>
          <span className={styles.city}>{flight.from.city}</span>
          <span className={styles.iata}>{flight.from.iata}</span>
        </div>
        <FlightStatus progress={progress} />
        <div className={`${styles.airport} ${styles.to}`}>
          <span className={styles.city}>{flight.to.city}</span>
          <span className={styles.iata}>{flight.to.iata}</span>
        </div>
      </div>
    </div>
  )
}