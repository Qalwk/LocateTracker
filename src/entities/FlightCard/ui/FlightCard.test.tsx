import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";

import { MemoryRouter } from "react-router";

import type { Flight } from "shared/mocks/FlightsData";

import { FlightCard } from "./FlightCard";

const mockFlight: Flight = {
  id: "S7129",
  airline: "S7 Airlines",
  airlineLogo: "images.png",
  flightNumber: "S7129",
  codes: ["88020", "RA-73420"],
  from: {
    city: "Moscow",
    iata: "DME",
    utc: "UTC+3",
    distance: 3200,
    duration: "4h 30m",
    scheduled: "22:00",
    actual: "22:10",
    countryCode: "RU",
  },
  to: {
    city: "Dubai",
    iata: "DXB",
    utc: "UTC+4",
    distance: 3200,
    duration: "4h 30m",
    scheduled: "02:30",
    estimated: "02:40",
    countryCode: "AE",
  },
  aircraft: "Airbus A321neo",
  speed: "860 km/h",
  altitude: "11 200 m",
};

test("отображает переданные данные рейса", () => {
  render(
    <MemoryRouter>
      <FlightCard
        flight={mockFlight}
        isFavorite={false}
        onLikeClick={() => {}}
        progress={60}
      />
    </MemoryRouter>,
  );

  // Проверяем alt атрибут изображения авиакомпании
  expect(screen.getByAltText("airlineLogo")).toBeInTheDocument();
  expect(screen.getByText("Moscow")).toBeInTheDocument();
  expect(screen.getByText("Dubai")).toBeInTheDocument();
  expect(screen.getByText("S7129")).toBeInTheDocument();
  expect(screen.getByText("DME")).toBeInTheDocument();
  expect(screen.getByText("DXB")).toBeInTheDocument();
});

test("вызывает onLikeClick при клике на кнопку лайка", async () => {
  const onLikeClick = jest.fn();

  render(
    <MemoryRouter>
      <FlightCard
        flight={mockFlight}
        isFavorite={false}
        onLikeClick={onLikeClick}
        progress={60}
      />
    </MemoryRouter>,
  );

  const likeButton = screen.getByRole("button");
  await userEvent.click(likeButton);

  expect(onLikeClick).toHaveBeenCalledTimes(1);
});

test("отображает активное состояние лайка", () => {
  render(
    <MemoryRouter>
      <FlightCard
        flight={mockFlight}
        isFavorite={true}
        onLikeClick={() => {}}
        progress={60}
      />
    </MemoryRouter>,
  );

  const likeButton = screen.getByRole("button");
  expect(likeButton).toHaveClass("btnLikeActive");
});

test("вызывает onClick при клике на карточку", async () => {
  const onClick = jest.fn();

  render(
    <MemoryRouter>
      <FlightCard
        flight={mockFlight}
        onClick={onClick}
        isFavorite={false}
        onLikeClick={() => {}}
        progress={60}
      />
    </MemoryRouter>,
  );

  const card = screen.getByText("Moscow").closest("div");
  await userEvent.click(card!);

  expect(onClick).toHaveBeenCalledWith(mockFlight);
});

test("соответствует снепшоту", () => {
  const { asFragment } = render(
    <MemoryRouter>
      <FlightCard
        flight={mockFlight}
        isFavorite={false}
        onLikeClick={() => {}}
        progress={60}
      />
    </MemoryRouter>,
  );
  expect(asFragment()).toMatchSnapshot();
});
