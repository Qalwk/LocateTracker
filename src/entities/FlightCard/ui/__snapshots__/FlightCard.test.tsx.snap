// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`соответствует снепшоту 1`] = `
<DocumentFragment>
  <div
    class="flightCard"
  >
    <div
      class="header"
    >
      <div
        class="airline"
      >
        <div
          class="airlineLogo"
        >
          <img
            alt="airlineLogo"
            src="http://localhost:3001/images/images.png"
          />
        </div>
        <div
          class="flightNumber"
        >
          S7129
        </div>
      </div>
      <div
        class="codes"
      >
        <span
          class="code"
        >
          88020
        </span>
        <span
          class="code codeSecond"
        >
          RA-73420
        </span>
        <button
          class="btnLike"
        >
          <svg
            aria-hidden="true"
            class="lucide lucide-heart"
            fill="none"
            height="16"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="16"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M19 14c1.49-1.46 3-3.21 3-5.5A5.5 5.5 0 0 0 16.5 3c-1.76 0-3 .5-4.5 2-1.5-1.5-2.74-2-4.5-2A5.5 5.5 0 0 0 2 8.5c0 2.3 1.5 4.05 3 5.5l7 7Z"
            />
          </svg>
        </button>
      </div>
    </div>
    <div
      class="airports"
    >
      <div
        class="airport from"
      >
        <span
          class="city"
        >
          Moscow
        </span>
        <span
          class="iata"
        >
          DME
        </span>
      </div>
      <div
        class="routeLine"
      >
        <div
          class="routeProgress"
          style="width: 60%;"
        />
        <span
          class="planeIcon"
          style="left: 60%;"
        >
          ✈️
        </span>
      </div>
      <div
        class="airport to"
      >
        <span
          class="city"
        >
          Dubai
        </span>
        <span
          class="iata"
        >
          DXB
        </span>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
